joomla/joomla-cms
# Add upload file size check to Joomla Update Component Upload & Update

Pull Request for Issue #27494 .

### Summary of Changes

This Pull Request (PR) is a back port of PR #27570 from 4.0-dev.

It adds following to the Upload & Update tab of the Joomla Update Component:
- Display of the selected upload file's size.
- Check if the selected file is too big to be uploaded with respect to the PHP limits.

If the selected upload file is too big, a warning alert is shown. If the user tries to upload that file anway, the upload is rejected with an appropriate browser alert message.

The size of the selected file is always shown in MB because the PHP limits and so the maximum file size will with a 99.9 % likelyhood also be in MB.

For the Upload & Install Package tab of the Extension Installer I've created PR #30190 for backporting PR #27665 from 4.0-dev.

### Testing Instructions

1. Go to `administrator/index.php?option=com_joomlaupdate` and there to tab "Upload & Update".

2. Try to upload a file which is bigger than the maximum size shown below the file select button.
**Hint:** This does not necessarily need to be a Joomla Update paclage. Take any file with a sufficient size.
**Result:** See section "Actual result BEFORE applying this Pull Request" below.

3. Apply patch of this PR.

4. Go to `administrator/index.php?option=com_joomlaupdate` and there to tab "Upload & Update".

5. Check the `Joomla package file` fields.
**Result:** See section "Expected result" screenshot 1. No information related to the selected file's size is shown if no file is selected.

6. Select a file which is **not** too big compared to the maximum upload size shown below the file selection.
**Hint:** This can be any file, it doesn't necessarily have to be a Joomla Update package.
**Result:** See section "Expected result" screenshot 2. The size of the selected file is shown below the maximum upload size.

7. Select a file which **is** too big compared to the maximum upload size shown below the file selection.
**Hint:** This can be any file, it doesn't necessarily have to be a Joomla Update package.
**Result:** See section "Expected result" screenshot 3. The size of the selected file is shown below the maximum upload size, and below that a warning alert is shown telling that the selected file is too big.

8. Try to upload the file anway by selecting the `Upload & Install` button.
**Result:** See section "Expected result" screenshot 4. A browser alert is shown telling that the file can't be uploaded because it's too big. The upload is not started.

9. Toggle the file selection a few times between too big and not too big file, and also between different files of each type.
**Result:** The warning disappears when the selection is changed from too big to not too big, and it appears again when vice versa. The file size is updated with each change of the file.

10. Finally select a valid Joomla Update package which is not too big and verify that uploading still works.
**Result:** Uploading of not too big file still works as well as without this PR.

### Actual result BEFORE applying this Pull Request

The result depends on the values of the PHP limit `post_max_size` and `upload_max_filesize`.

- If `post_max_size` is big enough but `upload_max_filesize` is violated, you get an error alert with the untranslated language string constants `COM_INSTALLER_MSG_INSTALL_WARNINSTALLUPLOADERROR` and `COM_INSTALLER_MSG_WARNINGS_SMALLUPLOADSIZE`.
The not translated languages strings will be solved with PR #29899 .

- If `upload_max_filesize` is big enough but `post_max_size` is violated, you get stuck in the update process, i.e. nothing happens, no progress.

### Expected result AFTER applying this Pull Request

- No file selected:
![expected-result_no-file-selected](https://user-images.githubusercontent.com/7413183/86342226-9bc86680-bc57-11ea-9313-22d29326f561.png)

- File size ok:
![expected-result_file-size-ok](https://user-images.githubusercontent.com/7413183/86342264-a551ce80-bc57-11ea-8012-a718e6fad676.png)

- File too big:
![expected-result_file-size-too-big](https://user-images.githubusercontent.com/7413183/86342274-a97dec00-bc57-11ea-8f82-08e800048865.png)

- Broswer alert when trying to upload too big file:
![expected-result_browser-alert](https://user-images.githubusercontent.com/7413183/86342288-ad117300-bc57-11ea-834c-05a7529e07ca.png)

### Documentation Changes Required

None.
